src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,63 @@
 package com.es.phoneshop.model.product;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private static ArrayListProductDao instance;
+    private volatile ArrayList<Product> products = new ArrayList<>();
+
+    private ArrayListProductDao() {
+    }
+
+    public static ArrayListProductDao getInstance() {
+        if (instance == null) {
+            synchronized (ArrayListProductDao.class) {
```

Comment:
You should avoid using class instances as locks. Rollback the singleton-related changes in scope of this task, there are some tricky notes about singletons which we will discuss on the next week